File: lib/shared/services/user_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../models/user_model.dart';
import '../../core/config/firebase_config.dart';

/// Service for managing user data
class UserService {
  final FirebaseFirestore? _firestore = FirebaseConfig.isAvailable
      ? FirebaseFirestore.instance
      : null;

  /// Get user stream
  Stream<UserModel?> getUser(String userId) {
    if (_firestore == null) {
      return Stream.value(null);
    }

    return _firestore!
        .collection('users')
        .doc(userId)
        .snapshots()
        .map((snapshot) {
      if (!snapshot.exists) return null;
      return UserModel.fromFirestore(snapshot);
    });
  }

  /// Get user once
  Future<UserModel?> getUserOnce(String userId) async {
    if (_firestore == null) return null;

    final snapshot = await _firestore!.collection('users').doc(userId).get();

    if (!snapshot.exists) return null;
    return UserModel.fromFirestore(snapshot);
  }

  /// Create or update user
  Future<void> saveUser(UserModel user) async {
    if (_firestore == null) return;

    await _firestore!
        .collection('users')
        .doc(user.id)
        .set(user.toFirestore(), SetOptions(merge: true));
  }

  /// Update user fields
  Future<void> updateUser(String userId, Map<String, dynamic> data) async {
    if (_firestore == null) return;

    await _firestore!.collection('users').doc(userId).update(data);
  }

  /// Update user points
  Future<void> updatePoints(String userId, int points) async {
    if (_firestore == null) return;

    await _firestore!.collection('users').doc(userId).update({
      'totalPoints': FieldValue.increment(points),
    });
  }

  /// Update user streak
  Future<void> updateStreak(String userId, int streak) async {
    if (_firestore == null) return;

    await _firestore!.collection('users').doc(userId).update({
      'dailyStreak': streak,
      'lastActiveDate': Timestamp.now(),
    });
  }

  /// Add badge to user
  Future<void> addBadge(String userId, String badgeId) async {
    if (_firestore == null) return;

    await _firestore!.collection('users').doc(userId).update({
      'badges': FieldValue.arrayUnion([badgeId]),
    });
  }

  /// Increment AI diagnosis count
  Future<void> incrementDiagnosisCount(String userId) async {
    if (_firestore == null) return;

    await _firestore!.collection('users').doc(userId).update({
      'aiDiagnosisUsed': FieldValue.increment(1),
    });
  }

  /// Reset monthly diagnosis count
  Future<void> resetMonthlyDiagnosisCount(String userId) async {
    if (_firestore == null) return;

    await _firestore!.collection('users').doc(userId).update({
      'aiDiagnosisUsed': 0,
    });
  }

  /// Upgrade to premium
  Future<void> upgradeToPremium(String userId) async {
    if (_firestore == null) return;

    await _firestore!.collection('users').doc(userId).update({
      'isPremium': true,
    });
  }
}

File: src/services/userService.ts
import { db } from './firebase';
import {
  doc,
  getDoc,
  setDoc,
  updateDoc,
  increment,
  arrayUnion,
  onSnapshot,
} from 'firebase/firestore';
import { User } from '../models/types';
import { COLLECTIONS } from '../constants/app';

export async function getOrCreateUser(
  uid: string,
  email: string,
  displayName: string,
): Promise<User> {
  const userRef = doc(db, COLLECTIONS.users, uid);
  const userSnap = await getDoc(userRef);

  if (userSnap.exists()) {
    return { id: userSnap.id, ...userSnap.data() } as User;
  }

  const newUser: Omit<User, 'id'> = {
    email,
    displayName,
    isPremium: false,
    totalPoints: 0,
    dailyStreak: 0,
    aiDiagnosisUsed: 0,
    badges: [],
    createdAt: new Date().toISOString(),
  };

  await setDoc(userRef, newUser);
  return { id: uid, ...newUser };
}

export async function addPoints(
  userId: string,
  points: number,
): Promise<void> {
  const userRef = doc(db, COLLECTIONS.users, userId);
  await updateDoc(userRef, {
    totalPoints: increment(points),
  });
}

export async function updateStreak(userId: string): Promise<void> {
  const userRef = doc(db, COLLECTIONS.users, userId);
  await updateDoc(userRef, {
    dailyStreak: increment(1),
  });
}

export async function addBadge(
  userId: string,
  badgeId: string,
): Promise<void> {
  const userRef = doc(db, COLLECTIONS.users, userId);
  await updateDoc(userRef, {
    badges: arrayUnion(badgeId),
  });
}

export function subscribeToUser(
  userId: string,
  callback: (user: User | null) => void,
): () => void {
  try {
    const userRef = doc(db, COLLECTIONS.users, userId);
    return onSnapshot(
      userRef,
      (snapshot) => {
        if (snapshot.exists()) {
          callback({ id: snapshot.id, ...snapshot.data() } as User);
        } else {
          callback(null);
        }
      },
      (error) => {
        console.error('Error subscribing to user:', error);
        callback(null);
      },
    );
  } catch (error) {
    console.error('Error setting up user subscription:', error);
    callback(null);
    return () => {};
  }
}
